### PR TITLE
Fix blocking callback null validation

### PR DESCRIPTION
**Pull request by vega113**
Refs #990

## Summary
- fix `BlockingSuccessFailCallback.onSuccess` so null success values are rejected
- replace the stale crash-vs-log TODO with an explicit fail-fast contract for invalid callback use
- use a single atomic completion gate so concurrent mixed callbacks cannot publish both success and failure
- add focused tests for success, failure, timeout, null callbacks, sequential duplicate/mixed callbacks, and concurrent mixed callbacks

## Verification
- `python3 scripts/assemble-changelog.py` -> assembled 233 entries into ignored local `wave/config/changelog.json`
- red check before fix: `sbt "testOnly org.waveprotocol.box.server.util.BlockingSuccessFailCallbackTest"` failed on null success validation as expected
- `sbt "testOnly org.waveprotocol.box.server.util.BlockingSuccessFailCallbackTest"` -> Passed 10 tests, 0 failed after the concurrency fix and rebase onto current `origin/main`
- `git diff --check HEAD~1..HEAD` -> passed
- `git status --short --branch` -> clean, ahead by this PR commit

## Reviews
- Internal spec review initially found the concurrent mixed-callback race; fixed with a single atomic terminal state
- Internal code-quality review initially found the same race; fixed
- Internal spec re-review: no findings
- Internal code-quality re-review: no findings; reran the focused callback test successfully

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved callback completion enforcement to prevent duplicate completion attempts and invalid state transitions.
  * Enhanced validation with immediate error feedback for incorrect usage patterns.

* **Tests**
  * Added comprehensive test coverage validating callback terminal semantics and input validation requirements.
  * Included concurrency testing to ensure correct behavior under concurrent access scenarios.

### CODE REVIEW

**Comment by vega113** [comment]
Applied `skip-changelog` after inspecting the changelog workflow earlier in this issue. This PR is internal utility cleanup plus focused tests for a blocking callback helper; it does not change user-facing behavior.

File: wave/src/main/java/org/waveprotocol/box/server/util/BlockingSuccessFailCallback.java [hunk 1]
```diff
@@ -86,17 +88,17 @@ public Pair<S, F> await(long timeout, TimeUnit unit) {
...
 
   @Override
   public void onSuccess(S successValue) {
     LOG.fine(description + ": onSuccess(" + successValue + ")");
-    Preconditions.checkArgument(successResult != null);
-    Preconditions.checkState(successResult.getAndSet(successValue) == null);
-    Preconditions.checkState(failureResult.get() == null);
+    Preconditions.checkArgument(successValue != null);
+    Preconditions.checkState(completed.compareAndSet(false, true));
+    successResult.set(successValue);
```
[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/main/java/org/waveprotocol/box/server/util/BlockingSuccessFailCallback.java]: Fixed in 2a9825f. Added descriptive messages to all Preconditions calls.

File: wave/src/test/java/org/waveprotocol/box/server/util/BlockingSuccessFailCallbackTest.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        } finally {
+          done.countDown();
+        }
+      }
+    });
+
+    successThread.start();
+    failureThread.start();
+    start.countDown();
+    assertTrue(done.await(1, TimeUnit.SECONDS));
```
[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/test/java/org/waveprotocol/box/server/util/BlockingSuccessFailCallbackTest.java]: Fixed in 2a9825f. Increased latch timeout from 1s to 5s and added an assertion message: `assertTrue("Timed out waiting for concurrent callbacks", done.await(5, TimeUnit.SECONDS))`.

File: wave/src/main/java/org/waveprotocol/box/server/util/BlockingSuccessFailCallback.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/main/java/org/waveprotocol/box/server/util/BlockingSuccessFailCallback.java]: Fixed in 2a9825f70: added descriptive precondition messages for null values and already-completed duplicate/mixed callbacks, including the callback description.

File: wave/src/main/java/org/waveprotocol/box/server/util/BlockingSuccessFailCallback.java [hunk 3]
```diff
@@ -86,17 +88,17 @@ public Pair<S, F> await(long timeout, TimeUnit unit) {
   public void onFailure(F failureValue) {
     LOG.warning(description + ": onFailure(" + failureValue + ")");
     Preconditions.checkArgument(failureValue != null);
-    Preconditions.checkState(failureResult.getAndSet(failureValue) == null);
-    Preconditions.checkState(successResult.get() == null);
+    Preconditions.checkState(completed.compareAndSet(false, true));
+    failureResult.set(failureValue);
     awaitLatch.countDown();
```
[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/main/java/org/waveprotocol/box/server/util/BlockingSuccessFailCallback.java:95]: Fixed in 2a9825f70: moved the warning log after validation and the completion gate, so rejected null/duplicate/mixed calls do not emit warning logs.

File: wave/src/test/java/org/waveprotocol/box/server/util/BlockingSuccessFailCallbackTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/test/java/org/waveprotocol/box/server/util/BlockingSuccessFailCallbackTest.java]: Fixed in 2a9825f70: increased the concurrent callback wait to 5 seconds and added an assertion message.

File: wave/src/main/java/org/waveprotocol/box/server/util/BlockingSuccessFailCallback.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/main/java/org/waveprotocol/box/server/util/BlockingSuccessFailCallback.java:95]: Fixed in 2a9825f. Moved LOG.warning in onFailure (and LOG.fine in onSuccess) to after the checkArgument/checkState guards, so invalid calls (null/duplicate) are rejected before any log entry is emitted.
